cloudinit/config/cc_growpart.py

```diff
@@ -274,12 +275,69 @@ def devent2dev(devent):
     return dev
 
 
+def is_mapped_device(blockdev) -> bool:
+    """
+    Check if a device is a mapped device.
+
+    blockdev should look something like '/dev/mapper/disk1' if True,
+    otherwise something like /dev/vdb1.
+    """
+    is_mapped = blockdev.startswith('/dev/mapper/')
+    LOG.debug("%s is %s a mapped device", blockdev, "" if is_mapped else "not")
+    return is_mapped
+
+
+def is_encrypted(blockdev) -> bool:
+    """
+    Check if a device is an encrypted device.
+
+    Will work for both mapped and raw devices.
+    """
+    is_encrypted = False
+    if not subp.which('cryptsetup'):
+        LOG.debug('cryptsetup not found. Assuming no encrypted partitions')
+        return False
+    with suppress(subp.ProcessExecutionError):
+        subp.subp(['cryptsetup', 'status', blockdev])
+        is_encrypted = True
+    if not is_encrypted:
+        with suppress(subp.ProcessExecutionError):
+            subp.subp(['cryptsetup', 'isLuks', blockdev])
+            is_encrypted = True
+    LOG.debug(
+        "Determined that %s is %s encrypted",
+        blockdev,
+        "" if is_encrypted else "not"
+    )
+    return is_encrypted
+
+
+def get_underlying_partition(blockdev):
+    command = 'dmsetup deps -o devname {}'.format(blockdev)
+    dep = subp.subp(command.split())[0]
+    try:
+        # Returned result should look something like:
+        # 1 dependencies : (vdb1)
+        return '/dev/{}'.format(dep.split(': (')[1].split(')')[0])
+    except IndexError:
+        raise Exception(
+            "Ran `{}`, but received unexpected stdout: `{}`".format(
+                command, dep))
+
+
+def resize_encrypted(blockdev):
+    subp.subp(['cryptsetup', '--key-file', '<keyfile>', 'resize', blockdev])
```

Comment:
This can't work as-is, but not sure where I'm supposed to get the key from.

Reply:
You would not need key-file at all, as it would be in kernel keyring already, because the drive is unlocked.

Reply:
However, there is backup recovery keyfile on disk too

Reply:
Ok, I pushed a change to remove that argument. It should be testable now.